CC-1291: Add `build-and-run` command (used in `check-go-is-present.yml`)

ryan-gang opened the pull request:
This pull request introduces a new feature to the command-line interface (CLI) of the application, specifically a new `check-go` command. This command checks if Go is present and configured correctly, which is essential for the shell challenge where Go is needed to build the custom executable. The changes involve the addition of a new `ValidateCommand` class and a `RuntimeTester` class.

Here are the most important changes:

New Command Addition:

* [`cli.ts`](diffhunk://#diff-34610464905445a5c4c85ac0f5f0c6e5c1491d074f895606359e1c5b82390cc7R46-R60): A new `check-go` command was added to the `program` object. This command takes an optional `language` argument and runs the `ValidateCommand` with the appropriate parameters.

New Classes:

* [`commands/validate.ts`](diffhunk://#diff-dcc0bcb05d2b5adb78f5b66ed793362e5c497d4b6e05825c66ab25a4888f444aR1-R52): The `ValidateCommand` class was introduced. It takes a `languageFilter`, `commandToExecute`, `outputStreamType`, and `expectedOutput` as parameters. The `doRun` method loads a `Course` from the current directory, filters the runtime testers based on the language filter, and runs the tests.
* [`lib/testers/runtime-tester.ts`](diffhunk://#diff-ce085e4bbb3208703d65268ad8d690bb7005d626013e6967a226305b0efb4148R1-R67): The `RuntimeTester` class was introduced. It takes a `Course`, `Dockerfile`, `commandToExecute`, `outputStreamType`, and `expectedOutput` as parameters. The `doTest` method builds an image and runs the specified command inside it, checking the output stream for the expected output.

Code Refactoring:

* [`cli.ts`](diffhunk://#diff-34610464905445a5c4c85ac0f5f0c6e5c1491d074f895606359e1c5b82390cc7L5-R5): The `Option` import was removed and replaced with a new `ValidateCommand` import from `./commands/validate`.

ryan-gang committed:
feat: add check-go cli command
added configurable runtime-tester class to run commands inside built docker images and assert output easily for all languages

ryan-gang committed:
feat: add verbose output from docker build for `check-go` command

ryan-gang committed:
feat: add logging for expected output in RuntimeTester

ryan-gang committed:
refactor: Use Promise.all to improve performance in runtimeTestersForCourse method

ryan-gang committed:
feat: Add GitHub Actions workflow to check if Go is present inside Docker container

ryan-gang commented:
The command in action: [build-your-own-shell#15](https://github.com/codecrafters-io/build-your-own-shell/actions/runs/9676428540/job/26695946799?pr=15)

ryan-gang committed:
feat: update interface to check-go-is-present ci test
Pass in params from CLI.

ryan-gang commented:
Updated the command name, and argument list. 
`course-sdk buildAndRun all "go version" "stdout" "go version"`

rohitpaulk commented on cli.ts:
```diff
@@ -43,4 +43,21 @@ program
   });
 
+program
+  .command("buildAndRun")
```
```suggestion
  .command("build-and-run")
```

(Seems more common to name commands this way rather than without dashes/underscores)

rohitpaulk commented on commands/validate.ts:
```diff
@@ -0,0 +1,52 @@
+import BaseCommand from "./base";
+import RuntimeTester from "../lib/testers/runtime-tester";
+
+export default class ValidateCommand extends BaseCommand {
```
Can we name this `BuildAndRun` too? Weird to have a command name that's different from the command class

rohitpaulk commented on lib/testers/runtime-tester.ts:
```diff
@@ -0,0 +1,72 @@
+
+const exec = util.promisify(child_process.exec);
+
+export default class RuntimeTester extends BaseTester {
```
```suggestion
export default class CommandTester extends BaseTester {
```

(Not super sure about this, just leaving as a suggestion)

ryan-gang committed:
refactor: Rename ValidateCommand to BuildAndRunCommand and update related classes

This commit renames the ValidateCommand class to BuildAndRunCommand to better reflect its purpose. The related classes and files have also been updated accordingly.

ryan-gang committed:
refactor: Update CLI command name to use kebab-case for consistency

The CLI command "buildAndRun" has been renamed to "build-and-run" to follow the kebab-case naming convention used throughout the codebase. This change improves consistency and readability.

ryan-gang committed:
chore: update workflow file to use update command name
